File: easy-random-bean-validation/src/test/java/org/jeasy/random/validation/NegativeOrZeroAnnotationHandlerTest.java
package org.jeasy.random.validation;

import jakarta.validation.ConstraintViolation;
import jakarta.validation.Validation;
import jakarta.validation.Validator;
import jakarta.validation.ValidatorFactory;
import jakarta.validation.constraints.NegativeOrZero;
import org.jeasy.random.EasyRandom;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.util.Set;

import static org.assertj.core.api.Assertions.assertThat;

class NegativeOrZeroAnnotationHandlerTest {
  private EasyRandom easyRandom;

  @BeforeEach
  void setUp() {
    easyRandom = new EasyRandom();
  }

  @Test
  void generatedBeanShouldBeValidAccordingToValidationConstraints() {
    byte byteValue = 0;
    short shortValue = 0;
    int intValue = 0;
    long longValue = 0L;
    float floatValue = 0F;

    TestBean testBean = easyRandom.nextObject(TestBean.class);

    assertThat(testBean.getTestBigDecimal()).isLessThanOrEqualTo(BigDecimal.ZERO);
    assertThat(testBean.getTestBigInteger()).isLessThanOrEqualTo(BigInteger.ZERO);
    assertThat(testBean.getTestPrimitiveByte()).isLessThanOrEqualTo(byteValue);
    assertThat(testBean.getTestWrapperByte()).isLessThanOrEqualTo(byteValue);
    assertThat(testBean.getTestPrimitiveShort()).isLessThanOrEqualTo(shortValue);
    assertThat(testBean.getTestWrapperShort()).isLessThanOrEqualTo(shortValue);
    assertThat(testBean.getTestPrimitiveInt()).isLessThanOrEqualTo(intValue);
    assertThat(testBean.getTestWrapperInteger()).isLessThanOrEqualTo(intValue);
    assertThat(testBean.getTestPrimitiveLong()).isLessThanOrEqualTo(longValue);
    assertThat(testBean.getTestWrapperLong()).isLessThanOrEqualTo(longValue);
    assertThat(testBean.getTestPrimitiveFloat()).isLessThanOrEqualTo(floatValue);
    assertThat(testBean.getTestWrapperFloat()).isLessThanOrEqualTo(floatValue);
    assertThat(testBean.getTestPrimitiveDouble()).isLessThanOrEqualTo(0D);
    assertThat(testBean.getTestWrapperDouble()).isLessThanOrEqualTo(0D);
  }

  @Test
  void generatedBeanShouldBeValidUsingBeanValidationApi() {
    TestBean testBean = easyRandom.nextObject(TestBean.class);

    Validator validator;
    try (ValidatorFactory validatorFactory = Validation.buildDefaultValidatorFactory()) {
      validator = validatorFactory.getValidator();
    }
    Set<ConstraintViolation<TestBean>> violations = validator.validate(testBean);

    assertThat(violations).isEmpty();
  }

  static class TestBean {
    public BigDecimal getTestBigDecimal() {
      return testBigDecimal;
    }

    public BigInteger getTestBigInteger() {
      return testBigInteger;
    }

    public byte getTestPrimitiveByte() {
      return testPrimitiveByte;
    }

    public Byte getTestWrapperByte() {
      return testWrapperByte;
    }

    public short getTestPrimitiveShort() {
      return testPrimitiveShort;
    }

    public Short getTestWrapperShort() {
      return testWrapperShort;
    }

    public int getTestPrimitiveInt() {
      return testPrimitiveInt;
    }

    public Integer getTestWrapperInteger() {
      return testWrapperInteger;
    }

    public long getTestPrimitiveLong() {
      return testPrimitiveLong;
    }

    public Long getTestWrapperLong() {
      return testWrapperLong;
    }

    public float getTestPrimitiveFloat() {
      return testPrimitiveFloat;
    }

    public Float getTestWrapperFloat() {
      return testWrapperFloat;
    }

    public double getTestPrimitiveDouble() {
      return testPrimitiveDouble;
    }

    public Double getTestWrapperDouble() {
      return testWrapperDouble;
    }

    @NegativeOrZero private BigDecimal testBigDecimal;

    @NegativeOrZero private BigInteger testBigInteger;

    @NegativeOrZero private byte testPrimitiveByte;

    @NegativeOrZero private Byte testWrapperByte;

    @NegativeOrZero private short testPrimitiveShort;

    @NegativeOrZero private Short testWrapperShort;

    @NegativeOrZero private int testPrimitiveInt;

    @NegativeOrZero private Integer testWrapperInteger;

    @NegativeOrZero private long testPrimitiveLong;

    @NegativeOrZero private Long testWrapperLong;

    @NegativeOrZero private float testPrimitiveFloat;

    @NegativeOrZero private Float testWrapperFloat;

    @NegativeOrZero private double testPrimitiveDouble;

    @NegativeOrZero private Double testWrapperDouble;
  }
}

File: bean-validation/src/test/kotlin/io/github/krandom/validation/NegativeOrZeroAnnotationHandlerTest.kt
package io.github.krandom.validation

import io.github.krandom.KRandom
import io.kotest.matchers.collections.shouldBeEmpty
import io.kotest.matchers.comparables.shouldBeLessThanOrEqualTo
import jakarta.validation.Validation
import jakarta.validation.Validator
import jakarta.validation.constraints.NegativeOrZero
import java.math.BigDecimal
import java.math.BigInteger
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

internal class NegativeOrZeroAnnotationHandlerTest {
  private lateinit var kRandom: KRandom

  @BeforeEach
  fun setUp() {
    kRandom = KRandom()
  }

  @Test
  fun generatedBeanShouldBeValidAccordingToValidationConstraints() {
    val byteValue: Byte = 0
    val shortValue: Short = 0
    val intValue = 0
    val longValue = 0L
    val floatValue = 0f
    val doubleValue = 0.0

    val testBean = kRandom.nextObject(TestBean::class.java)

    testBean.testBigDecimal shouldBeLessThanOrEqualTo BigDecimal.ZERO
    testBean.testBigInteger shouldBeLessThanOrEqualTo BigInteger.ZERO
    testBean.testPrimitiveByte shouldBeLessThanOrEqualTo byteValue
    testBean.testWrapperByte shouldBeLessThanOrEqualTo byteValue
    testBean.testPrimitiveShort shouldBeLessThanOrEqualTo shortValue
    testBean.testWrapperShort shouldBeLessThanOrEqualTo shortValue
    testBean.testPrimitiveInt shouldBeLessThanOrEqualTo intValue
    testBean.testWrapperInteger shouldBeLessThanOrEqualTo intValue
    testBean.testPrimitiveLong shouldBeLessThanOrEqualTo longValue
    testBean.testWrapperLong shouldBeLessThanOrEqualTo longValue
    testBean.testPrimitiveFloat shouldBeLessThanOrEqualTo floatValue
    testBean.testWrapperFloat shouldBeLessThanOrEqualTo floatValue
    testBean.testPrimitiveDouble shouldBeLessThanOrEqualTo doubleValue
    testBean.testWrapperDouble shouldBeLessThanOrEqualTo doubleValue
  }

  @Test
  fun generatedBeanShouldBeValidUsingBeanValidationApi() {
    val testBean = kRandom.nextObject(TestBean::class.java)

    val validator: Validator
    Validation.buildDefaultValidatorFactory().use { validatorFactory ->
      validator = validatorFactory.validator
    }
    val violations = validator.validate(testBean)

    violations.shouldBeEmpty()
  }

  internal data class TestBean(
    @field:NegativeOrZero val testBigDecimal: BigDecimal,
    @field:NegativeOrZero val testBigInteger: BigInteger,
    @field:NegativeOrZero val testPrimitiveByte: Byte,
    @field:NegativeOrZero val testWrapperByte: Byte,
    @field:NegativeOrZero val testPrimitiveShort: Short,
    @field:NegativeOrZero val testWrapperShort: Short,
    @field:NegativeOrZero val testPrimitiveInt: Int,
    @field:NegativeOrZero val testWrapperInteger: Int,
    @field:NegativeOrZero val testPrimitiveLong: Long,
    @field:NegativeOrZero val testWrapperLong: Long,
    @field:NegativeOrZero val testPrimitiveFloat: Float,
    @field:NegativeOrZero val testWrapperFloat: Float,
    @field:NegativeOrZero val testPrimitiveDouble: Double,
    @field:NegativeOrZero val testWrapperDouble: Double,
  )
}
